cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,880 @@
+		if err := transport.Serve(ctx, addr, as.cert, as); err != nil {
+			// only propagate non-context.Canceled errors
+			if !errors.Is(err, context.Canceled) {
+				as.debugf("server error: %v", err)
+				cancel()
+				serverError <- err
+			}
+		}
+	}()
+
+	// start periodic discovery of peers
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		periodic(ctx, time.Second*5, time.Second*1, func(ctx context.Context) {
+			discoveries, err := discover(ctx)
+			if err != nil {
+				as.debugf("error discovering peers: %v", err)
+				return
+			}
+
+			// filter out our address
+			addrs := make([]string, 0, len(discoveries))
+			for _, d := range discoveries {
+				if d == addr {
+					continue
+				}
+				addrs = append(addrs, d)
+			}
+
+			if err := as.publishAuth(ctx, addrs, client); err != nil {
+				as.debugf("error publishing auth messages: %v", err)
+				return
+			}
+		})
+	}()
+
+	var rounds int
+
+	// start up the periodic publication of routes
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		const (
+			period = time.Second * 5
+			jitter = time.Second
+			peers  = 5
+			routes = 5000
+		)
+		periodic(ctx, period, jitter, func(ctx context.Context) {
+			as.publishRoutes(ctx, client, peers, routes)
+			rounds++
+		})
+	}()
+
+	// start event-driven device operations
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for {
+			select {
+			case <-as.devices.PendingResponses():
+				as.publishDevices(ctx, client)
+			case <-as.devices.PendingOutgoingQueries():
+				as.publishDeviceQueries(ctx, client)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// wait for context cancellation
+	wg.Wait()
+
+	select {
+	case err := <-serverError:
+		return Routes{}, fmt.Errorf("server failed: %w", err)
+	default:
+	}
+
+	// perform final fingerprint consistency check
+	devices := as.devices.Export()
+	for _, identity := range devices.IDs {
+		if fp, ok := as.fingerprints[identity.RDT]; ok {
+			if fp != identity.FP {
+				return Routes{}, fmt.Errorf("consistency check failed: fingerprint mismatch for device %s", identity.RDT)
+			}
+		}
+	}
+
+	sent, received, tx, rx := transport.Stats()
+	as.debugf(
+		"assemble stopped after %d rounds, sent: %d messages (%d bytes), received: %d messages (%d bytes)",
+		rounds, sent, tx, received, rx,
+	)
+
+	return as.selector.Routes(), nil
+}
+
+func (as *AssembleState) debugf(format string, a ...any) {
+	as.logger.Debug(fmt.Sprintf(format, a...))
+}
+
+func periodic(
+	ctx context.Context,
+	interval time.Duration,
+	jitter time.Duration,
+	work func(ctx context.Context),
+) {
+	delay := func() time.Duration {
+		if jitter <= 0 {
+			return interval
+		}
+
+		// +- jitter from the given interval
+		j := time.Duration(randutil.Int63n(int64(jitter)*2)) - jitter
+		return interval + j
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(delay()):
+		}
+
+		// even if the timer won the select, we should still check if the
+		// context has been cancelled
+		if ctx.Err() != nil {
```

Comment:
Also the tests never get past this check so there's some logic in Run that doesn't get tested